### ISSUE TITLE
pytorch model: Failed to retrieve a prediction response within the specified latency SLO

### ISSUE THREAD

**Issue by zoux86** [ucbrise/clipper#563]
This is my code:

from clipper_admin import ClipperConnection, KubernetesContainerManager
clipper_conn = ClipperConnection(KubernetesContainerManager(useInternalIP=True))
clipper_conn.connect()
from torch import nn
import torch
model = nn.Linear(1, 1)
def predict(model,inputs):
    print(inputs)
    conv = torch.autograd.Variable(torch.randn(1, 1))
    pred = model(conv)
    pred = pred.data.numpy()
    return [str(x) for x in pred]
from clipper_admin.deployers.pytorch import create_endpoint
create_endpoint(clipper_conn, name="pytorch-1", input_type="int", func=predict, pytorch_model=model,version=1,slo_micros=3000000,pkgs_to_install=['--default-timeout=100 --upgrade pip','http://download.pytorch.org/whl/cpu/torch-0.4.1-cp27-cp27mu-linux_x86_64.whl','torchvision')

everything is ok, and the model is  linked to the application。the pods is running.
but when i  Query Clipper for predictions. it always show the message!
![image](https://user-images.githubusercontent.com/14087435/44825074-7d4c4b00-ac3b-11e8-88bf-aa6efbf78cf6.png)

i think there are some mistakes when i build pytorch model, can someone give me a suggestion or give me a example how to build pytorch model?

**Comment by withsmilo** [comment]
@zoux86 : Please check QueryFrontend's logs and `pytorch-1` model's logs.

**Comment by zoux86** [comment]
@withsmilo Thank you for your advice!
i check the logs, but i still don't know what's wrong! 
![image](https://user-images.githubusercontent.com/14087435/44827777-3d8b6080-ac47-11e8-8eff-47f1980e1f24.png)

here is the queryFronted's docker logs

2018-08-30T03:07:52.364355115Z [03:07:52.306][info]  [QUERYPR...] Found 1 tasks
2018-08-30T03:07:52.364396712Z [03:07:52.306][info]  [TASKEXE...] Adding task to queue. QueryID: 32, model: pytorch-zoux2:1
2018-08-30T03:07:52.364420965Z [03:07:52.307][info]  [TASKEXE...] Sending batch to model: pytorch-zoux2:1 replica 3.Batch size: 2. Query IDs: 32 32 
2018-08-30T03:07:53.171974096Z [03:07:52.977][info]         [RPC] Found message to receive
2018-08-30T03:07:53.172005126Z [03:07:53.020][info]         [RPC] Found message to receive
2018-08-30T03:07:53.344298926Z [03:07:53.323][info]         [RPC] Found message to receive
2018-08-30T03:07:54.356132749Z [03:07:54.175][info]         [RPC] Found message to receive
2018-08-30T03:07:54.758233566Z [03:07:54.694][info]         [RPC] Found message to receive
2018-08-30T03:07:54.758286444Z [03:07:54.703][info]         [RPC] Found message to receive
2018-08-30T03:07:55.168037987Z [03:07:54.998][info]         [RPC] Found message to receive
2018-08-30T03:07:55.168069297Z [03:07:55.000][info]         [RPC] Found message to receive
2018-08-30T03:07:55.168077434Z [03:07:55.000][info]         [RPC] New container connected
2018-08-30T03:07:55.168083498Z [03:07:55.000][info]         [RPC] Container added
2018-08-30T03:07:55.168089275Z [03:07:55.005][info]       [REDIS] Successfully issued command "SELECT 3"
2018-08-30T03:07:55.168095409Z [03:07:55.006][info]       [REDIS] Successfully issued command "HMSET pytorch-zoux2,1,4 model_id pytorch-zoux2:1 model_name pytorch-zoux2 model_version 1 model_replica_id 4 zmq_connection_id 39 batch_size 1 input_type integers"
2018-08-30T03:07:55.168102490Z [03:07:55.006][info]       [REDIS] MESSAGE: hset
2018-08-30T03:07:55.168108153Z [03:07:55.006][info]  [THREADPOOL] Work queue created for model pytorch-zoux2:1, replica 4
2018-08-30T03:07:55.168113759Z [03:07:55.010][info]       [REDIS] Successfully issued command "SELECT 3"
2018-08-30T03:07:55.168119482Z [03:07:55.010][info]       [REDIS] Successfully issued command "HGETALL pytorch-zoux2,1,4"
2018-08-30T03:07:55.168125283Z [03:07:55.010][info]       [REDIS] 	 model_id: pytorch-zoux2:1
2018-08-30T03:07:55.168131037Z [03:07:55.010][info]       [REDIS] 	 model_name: pytorch-zoux2
2018-08-30T03:07:55.168136683Z [03:07:55.010][info]       [REDIS] 	 model_version: 1
2018-08-30T03:07:55.168142286Z [03:07:55.010][info]       [REDIS] 	 model_replica_id: 4
2018-08-30T03:07:55.168147876Z [03:07:55.010][info]       [REDIS] 	 zmq_connection_id: 39
2018-08-30T03:07:55.168153574Z [03:07:55.010][info]       [REDIS] 	 batch_size: 1
2018-08-30T03:07:55.168186489Z [03:07:55.010][info]       [REDIS] 	 input_type: integers
2018-08-30T03:07:55.168196717Z [03:07:55.010][info]  [CONTAINERS] Adding new container - model: pytorch-zoux2, version: 1, connection ID: 39, replica ID: 4, input_type: integers
2018-08-30T03:07:55.168203215Z [03:07:55.010][info]  [CONTAINERS] Creating new ModelContainer for model pytorch-zoux2:1, id: 39
2018-08-30T03:07:55.168208921Z [03:07:55.011][info]  [CONTAINERS] 
2018-08-30T03:07:55.168214323Z Active containers:
2018-08-30T03:07:55.168219345Z 	Model: pytorch-zoux2:1
2018-08-30T03:07:55.168224688Z 		rep_id: 3, container_id: 38
2018-08-30T03:07:55.168230334Z 		rep_id: 4, container_id: 39
2018-08-30T03:07:55.168235731Z 	Model: pytorch-zoux1:1
2018-08-30T03:07:55.168240825Z 		rep_id: 3, container_id: 34
2018-08-30T03:07:55.168245930Z 	Model: pytorch-model-9:1
2018-08-30T03:07:55.168251249Z 		rep_id: 2, container_id: 28
2018-08-30T03:07:55.168256434Z 	Model: boston-model-2:1
2018-08-30T03:07:55.168261588Z 		rep_id: 2, container_id: 33
2018-08-30T03:07:55.168266816Z 	Model: pytorch-model-1:1
2018-08-30T03:07:55.168272021Z 		rep_id: 4, container_id: 7
2018-08-30T03:07:55.168277230Z 	Model: scipy-2:1
2018-08-30T03:07:55.168284089Z 		rep_id: 0, container_id: 1
2018-08-30T03:07:55.168289576Z 	Model: model-1:1
2018-08-30T03:07:55.168294741Z 		rep_id: 0, container_id: 0
2018-08-30T03:07:55.168299935Z 	Model: pytorch-model-3:1
2018-08-30T03:07:55.168305130Z 		rep_id: 1, container_id: 12
2018-08-30T03:07:55.168310422Z 	Model: pytorch-model-2:1
2018-08-30T03:07:55.168315467Z 		rep_id: 2, container_id: 10
2018-08-30T03:07:55.168320712Z 	Model: pytorch-model-8:1
2018-08-30T03:07:55.168325796Z 		rep_id: 2, container_id: 24
2018-08-30T03:07:55.168330953Z 	Model: pytorch-model-5:1
2018-08-30T03:07:55.168336015Z 		rep_id: 3, container_id: 25
2018-08-30T03:07:55.168341132Z 	Model: pytorch-model-6:1
2018-08-30T03:07:55.168346206Z 		rep_id: 1, container_id: 17
2018-08-30T03:07:55.168351318Z 	Model: pytorch-model-7:1
2018-08-30T03:07:55.168356372Z 		rep_id: 3, container_id: 21
2018-08-30T03:07:55.168361545Z 
2018-08-30T03:07:55.168366426Z [03:07:55.014][info]       [REDIS] Successfully issued command "SELECT 2"
2018-08-30T03:07:55.168372104Z [03:07:55.015][info]       [REDIS] Successfully issued command "HGETALL pytorch-zoux2:1"
2018-08-30T03:07:55.168377855Z [03:07:55.015][info]       [REDIS] 	 model_name: pytorch-zoux2
2018-08-30T03:07:55.168387141Z [03:07:55.015][info]       [REDIS] 	 model_version: 1
2018-08-30T03:07:55.168392912Z [03:07:55.015][info]       [REDIS] 	 load: 0.000000
2018-08-30T03:07:55.168398402Z [03:07:55.015][info]       [REDIS] 	 input_type: integers
2018-08-30T03:07:55.168403856Z [03:07:55.015][info]       [REDIS] 	 labels: 
2018-08-30T03:07:55.168409366Z [03:07:55.015][info]       [REDIS] 	 container_name: 192.168.13.87:5000/pytorch-zoux2:1
2018-08-30T03:07:55.168415012Z [03:07:55.015][info]       [REDIS] 	 model_data_path: DEPRECATED
2018-08-30T03:07:55.168420539Z [03:07:55.015][info]       [REDIS] 	 batch_size: -1
2018-08-30T03:07:55.168425939Z [03:07:55.038][info]  [THREADPOOL] Work queue created for model pytorch-zoux2:1, replica 4
2018-08-30T03:07:55.168431517Z [03:07:55.049][error] [THREADPOOL] Work queue already exists for model pytorch-zoux2:1, replica 4
2018-08-30T03:07:55.250481662Z [03:07:55.240][info]         [RPC] Found message to receive
2018-08-30T03:07:55.250500456Z [03:07:55.240][info]         [RPC] Found message to receive
2018-08-30T03:07:55.250504789Z [03:07:55.240][info]         [RPC] Found message to receive
2018-08-30T03:07:56.453103604Z [03:07:56.431][info]         [RPC] Found message to receive
2018-08-30T03:07:56.855618440Z [03:07:56.825][info]         [RPC] Found message to receive
2018-08-30T03:07:56.902968789Z [03:07:56.895][info]         [RPC] Found message to receive


and here is the pytorch model's container logs.
![image](https://user-images.githubusercontent.com/14087435/44827703-da013300-ac46-11e8-8661-f95685027cea.png)

**Comment by withsmilo** [comment]
@zoux86 : 
Your model's name in your shared codes is `pytorch-1`, but you queried to the `pytorch-zoux2` model.

**Comment by zoux86** [comment]
@withsmilo 
I build three apps, and three modes pytorch-1, pytorch-2,pytorch-zoux2, they all successed.
but when i query these models, it always show the default result。
i am wondering that there are some mistakes in my code ?

**Comment by withsmilo** [comment]
@zoux86 : 
hmm... I didn't know why your models do not operate because I have not used for Pytorch deployment. How about refer to [this code](https://github.com/ucbrise/clipper/blob/develop/integration-tests/deploy_pytorch_models.py)?

**Comment by RehanSD** [comment]
Hi @zoux86 was this issue resolved? If not, would you mind telling me what pytorch version you are running locally? Currently, the 0.4.1 release uses a serialization method that is not supported by the 0.4.0 version running on the deployment container.

**Comment by rkooo567** [comment]
Closed as serialization is resolved.
